# Assignment 2

This is assignment 2 for the Data Management and Visualisation Course from the University of Wesleyan and Coursera

In this assignment we needed to load our chosen data set and produce some frequency distributions.

In [1]:
# importing the libraries I need
import pandas as pd
import numpy as np
import urllib.request

In [2]:
# downloading the data from Our World in Data and saving it locally
url = 'https://raw.githubusercontent.com/owid/covid-19-data/master/public/data/owid-covid-data.csv'
urllib.request.urlretrieve(url, 'data/covid_data.csv')

('data/covid_data.csv', <http.client.HTTPMessage at 0x19cba6a3940>)

In [3]:
# loading the data from the local file, and having a look at the first few rows
df = pd.read_csv('data/covid_data.csv')
df.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-02-24          1.0        1.0   
1      AFG      Asia  Afghanistan  2020-02-25          1.0        0.0   
2      AFG      Asia  Afghanistan  2020-02-26          1.0        0.0   
3      AFG      Asia  Afghanistan  2020-02-27          1.0        0.0   
4      AFG      Asia  Afghanistan  2020-02-28          1.0        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         NaN                  NaN  ...   
1                 NaN           NaN         NaN                  NaN  ...   
2                 NaN           NaN         NaN                  NaN  ...   
3                 NaN           NaN         NaN                  NaN  ...   
4                 NaN           NaN         NaN                  NaN  ...   

   gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
0        1803.987              NaN                597.029   
1        1803.987              NaN                597.029   
2        1803.987              NaN                597.029   
3        1803.987              NaN                597.029   
4        1803.987              NaN                597.029   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59             NaN           NaN                  37.746   
1                 9.59             NaN           NaN                  37.746   
2                 9.59             NaN           NaN                  37.746   
3                 9.59             NaN           NaN                  37.746   
4                 9.59             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  
0                         0.5            64.83                    0.511  
1                         0.5            64.83                    0.511  
2                         0.5            64.83                    0.511  
3                         0.5            64.83                    0.511  
4                         0.5            64.83                    0.511  

[5 rows x 59 columns]

Because for this assignment I want to look at the total case and death rates, it makes sense to only look at the data for the last date I have full data for. At the time of writing, the last date I had full data for was 2021-04-28, so I will limit my dataframe to that date

In [4]:
df.date = pd.to_datetime(df.date)
df_latest = df[df.date == '2021-04-28']  # last date we have data for
df_latest = df_latest.dropna(subset=['continent'])  # gets rid of summaries for 'world' and 'africa' etc, as I only want data for countries
df_latest

iso_code      continent     location       date  total_cases  new_cases  \
429        AFG           Asia  Afghanistan 2021-04-28      59370.0      145.0   
1301       ALB         Europe      Albania 2021-04-28     130859.0      123.0   
1731       DZA         Africa      Algeria 2021-04-28     121580.0      236.0   
2155       AND         Europe      Andorra 2021-04-28      13148.0       27.0   
2561       AGO         Africa       Angola 2021-04-28      26168.0      226.0   
...        ...            ...          ...        ...          ...        ...   
82835      VEN  South America    Venezuela 2021-04-28     194959.0     1238.0   
83298      VNM           Asia      Vietnam 2021-04-28       2865.0        8.0   
84147      YEM           Asia        Yemen 2021-04-28       6263.0       43.0   
84555      ZMB         Africa       Zambia 2021-04-28      91484.0       66.0   
84961      ZWE         Africa     Zimbabwe 2021-04-28      38191.0       27.0   

       new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
429               165.143        2611.0        13.0                7.714  ...   
1301              125.571        2386.0         3.0                4.000  ...   
1731              200.857        3234.0         9.0                8.857  ...   
2155               33.000         125.0         1.0                0.286  ...   
2561              183.571         591.0         4.0                3.000  ...   
...                   ...           ...         ...                  ...  ...   
82835            1173.429        2099.0        17.0               19.143  ...   
83298               7.571          35.0         0.0                0.000  ...   
84147              43.286        1216.0         9.0                9.857  ...   
84555              52.143        1249.0         0.0                1.571  ...   
84961              30.143        1565.0         0.0                1.429  ...   

       gdp_per_capita  extreme_poverty  cardiovasc_death_rate  \
429          1803.987              NaN                597.029   
1301        11803.431              1.1                304.195   
1731        13913.839              0.5                278.364   
2155              NaN              NaN                109.135   
2561         5819.495              NaN                276.045   
...               ...              ...                    ...   
82835       16745.022              NaN                204.850   
83298        6171.884              2.0                245.465   
84147        1479.147             18.8                495.003   
84555        3689.251             57.5                234.499   
84961        1899.775             21.4                307.846   

       diabetes_prevalence  female_smokers  male_smokers  \
429                   9.59             NaN           NaN   
1301                 10.08             7.1          51.2   
1731                  6.73             0.7          30.4   
2155                  7.97            29.0          37.8   
2561                  3.94             NaN           NaN   
...                    ...             ...           ...   
82835                 6.47             NaN           NaN   
83298                 6.00             1.0          45.9   
84147                 5.35             7.6          29.2   
84555                 3.94             3.1          24.7   
84961                 1.82             1.6          30.7   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
429                    37.746                        0.50            64.83   
1301                      NaN                        2.89            78.57   
1731                   83.741                        1.90            76.88   
2155                      NaN                         NaN            83.73   
2561                   26.664                         NaN            61.15   
...                       ...                         ...              ...   
82835       

The columns I want to make frequency tables for are 'total_cases_per_million', 'total_deaths_per_million' and 'human_development_index', so I will check for missing values from those columns.

In [5]:
print(df_latest.total_cases_per_million.isnull().value_counts())
print(df_latest.total_deaths_per_million.isnull().value_counts())
print(df_latest.human_development_index.isnull().value_counts())

False    190
True       8
Name: total_cases_per_million, dtype: int64
False    182
True      16
Name: total_deaths_per_million, dtype: int64
False    185
True      13
Name: human_development_index, dtype: int64


There aren't too many missing data, so for now I will restrict the dataframe to those columns, and drop all rows with missing data.

In [6]:
cols = ['location', 'total_cases_per_million', 'total_deaths_per_million', 'human_development_index']
df_latest = df_latest[cols].dropna(subset=cols)
df_latest

location  total_cases_per_million  total_deaths_per_million  \
429    Afghanistan                 1525.110                    67.072   
1301       Albania                45471.888                   829.106   
1731       Algeria                 2772.568                    73.750   
2155       Andorra               170167.605                  1617.809   
2561        Angola                  796.196                    17.982   
...            ...                      ...                       ...   
82835    Venezuela                 6856.076                    73.815   
83298      Vietnam                   29.433                     0.360   
84147        Yemen                  209.985                    40.770   
84555       Zambia                 4976.296                    67.940   
84961     Zimbabwe                 2569.548                   105.296   

       human_development_index  
429                      0.511  
1301                     0.795  
1731                     0.748  
2155                     0.868  
2561                     0.581  
...                        ...  
82835                    0.711  
83298                    0.704  
84147                    0.470  
84555                    0.584  
84961                    0.571  

[177 rows x 4 columns]

Because I am dealing with quantitative data here, not categorical data, doing a frequency table makes no sense, so I will first bin the data into categories.

In [7]:
df_latest['total_case_rate'] = pd.cut(df_latest.total_cases_per_million, 3, labels=['low', 'medium', 'high'])
df_latest['total_death_rate'] = pd.cut(df_latest.total_deaths_per_million, 3, labels=['low', 'medium', 'high'])
bins = np.linspace(0.350, 1.0, 6)  # creating 5 equally spaced bins
df_latest['development_rating'] = pd.cut(df_latest.human_development_index, bins, labels=['very low', 'low', 'medium', 'high', 'very high'])
df_latest

location  total_cases_per_million  total_deaths_per_million  \
429    Afghanistan                 1525.110                    67.072   
1301       Albania                45471.888                   829.106   
1731       Algeria                 2772.568                    73.750   
2155       Andorra               170167.605                  1617.809   
2561        Angola                  796.196                    17.982   
...            ...                      ...                       ...   
82835    Venezuela                 6856.076                    73.815   
83298      Vietnam                   29.433                     0.360   
84147        Yemen                  209.985                    40.770   
84555       Zambia                 4976.296                    67.940   
84961     Zimbabwe                 2569.548                   105.296   

       human_development_index total_case_rate total_death_rate  \
429                      0.511             low              low   
1301                     0.795             low              low   
1731                     0.748             low              low   
2155                     0.868            high           medium   
2561                     0.581             low              low   
...                        ...             ...              ...   
82835                    0.711             low              low   
83298                    0.704             low              low   
84147                    0.470             low              low   
84555                    0.584             low              low   
84961                    0.571             low              low   

      development_rating  
429                  low  
1301                high  
1731                high  
2155                high  
2561                 low  
...                  ...  
82835             medium  
83298             medium  
84147           very low  
84555                low  
84961                low  

[177 rows x 7 columns]

I can now do the frequency tables on these columns

In [8]:
df_latest.total_case_rate.value_counts(sort=False, normalize=True)

low       0.751412
medium    0.225989
high      0.022599
Name: total_case_rate, dtype: float64

In [9]:
df_latest.total_death_rate.value_counts(sort=False, normalize=True)

low       0.751412
medium    0.186441
high      0.062147
Name: total_death_rate, dtype: float64

In [10]:
df_latest.development_rating.value_counts(sort=False, normalize=True)

very low     0.084746
low          0.192090
medium       0.214689
high         0.316384
very high    0.192090
Name: development_rating, dtype: float64

## Summary

### Missing data
This was removed before I did the frequency tables, there were 21 rows removed altogether for missing data

### Total case rates and total death rates
Around three quarters of all countries are in the low category for total cases per million people.
Around three quarters of all countries are in the low category for total deaths per million people.
From these frequency tables alone it isn't possible to see how many are in the low category for both.
The majority of the rest of the countries are in the medium category for total cases total deaths, with very few being in the high category, although more (6%) are in the high category for deaths than for cases (2%).

### Human Development Index
This is more spaced out, as it has more categories. The biggest category is high with 32% of countries, the smallest is very low with 8% of countries. The other categories all contain approximately 20% of countries each.